File: backend/app/schemas/market.py
from datetime import datetime

from pydantic import BaseModel, Field

from decimal import Decimal

from app.schemas.base import MoneyField, NonNegativeMoney


class OutcomeResponse(BaseModel):
    id: str
    name: str
    outcome_index: int

    model_config = {"from_attributes": True}


class MarketResponse(BaseModel):
    id: str
    slug: str
    question: str
    description: str | None
    category: str | None
    status: str
    total_liquidity: MoneyField
    total_volume: MoneyField
    yes_price: MoneyField
    no_price: MoneyField
    closes_at: datetime
    winning_outcome_id: str | None = None
    winning_outcome_name: str | None = None
    outcomes: list[OutcomeResponse] | None = None

    model_config = {"from_attributes": True}


class MarketDetailResponse(MarketResponse):
    outcomes: list[OutcomeResponse]
    spread: MoneyField
    created_at: datetime


class MarketOutcomeCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    outcome_index: int = Field(..., ge=0)


class CreateMarketRequest(BaseModel):
    question: str = Field(..., min_length=5, max_length=1000)
    description: str | None = Field(None, max_length=5000)
    category: str | None = Field(None, max_length=100)
    slug: str = Field(..., min_length=3, max_length=255)
    closes_at: datetime
    initial_liquidity: NonNegativeMoney = Field(default=Decimal("0"))
    initial_probability: NonNegativeMoney | None = Field(default=None, ge=0.01, le=0.99)
    outcomes_create: list[MarketOutcomeCreate] | None = None


class MarketListResponse(BaseModel):
    success: bool = True
    data: list[MarketResponse]
    page: int
    page_size: int
    has_more: bool


class ResolveMarketRequest(BaseModel):
    winning_outcome_id: str

File: frontend/apps/web/hooks/api/types/market.ts
export interface MarketResponse {
  id: string
  slug: string
  question: string
  description: string | null
  category: string | null
  status: string
  total_liquidity: string
  total_volume: string
  yes_price: string
  no_price: string
  closes_at: string
  winning_outcome_id?: string
  winning_outcome_name?: string
  outcomes?: Outcome[]
}

export interface MarketListResponse {
  success: boolean
  data: MarketResponse[]
  page: number
  page_size: number
  has_more: boolean
}

export interface Outcome {
  id: string
  name: string
  outcome_index: number
}

export interface FAQ {
  id: string
  question: string
  answer: string
  display_order: number
}

export interface PriceHistoryPoint {
  timestamp: string
  outcomes: { id: string; name: string; price: string }[]
  total_volume: string
}

export interface MarketDetailResponse extends Omit<MarketResponse, "id"> {
  id: string
  outcomes: Outcome[]
  faqs?: FAQ[]
  spread: string
  created_at: string
}

export interface Trade {
  id: string
  market_id: string
  market_slug: string
  market_question: string
  outcome: string
  side: string
  price: string
  amount: string
  executed_at: string
  username: string
  total?: string
}

export interface TradesResponse {
  success: boolean
  data: {
    trades: Trade[]
    page: number
    page_size: number
  }
}

export interface MarketTrade {
  id: string
  outcome: string
  side: string
  price: string
  amount: string
  timestamp: string
  username: string
}

export interface MarketStats {
  total_volume: string
  total_liquidity: string
  num_trades: number
  yes_price: string
  no_price: string
  spread: string
  yes_liquidity: string
  no_liquidity: string
  status: string
}

export interface Holder {
  user_id: string
  username: string
  shares_held: string
  average_price: string
  realized_pnl: string
}

export interface CommentActivity {
  id: string
  user_id: string
  username: string
  content: string
  depth: number
  created_at: string
}

export interface MarketActivity {
  market_stats: MarketStats
  top_holders_by_outcome: Record<string, Holder[]>
  recent_trades: MarketTrade[]
  recent_comments: CommentActivity[]
}
